### ISSUE TITLE
r3000 MIPS output issues

### ISSUE THREAD

**Issue by mike-athene** [avast-tl/retdec#32]
Hi folks,

first of all, thank you very much for open-sourcing, I've been looking for a decent MIPS decompiler for a long time. 

I work on really old binaries (GCC 2.X era), so not sure if reports like this would be usefull or even apply to modern MIPS code, therefore any code pointers that may help me to debug on my own would be appreciated too.

[asm example](https://onlinedisassembler.com/odaweb/uAvx0qFn/0)

`bash .\bin\decompile.sh .\mips\ex1.bin -a mips -m raw -e little --raw-entry-point 0x8002DFF0 --raw-section-vma 0x8002DFF0`

```
int32_t entry_point(int32_t a1, int32_t a2) {
    *(int32_t *)-0x7ffaff04 = 0x10000 * a2 - 0x10000;
    *(int32_t *)-0x7ffaff08 = a1;
    return 0;
}
```

1.  Minor, but annoying: Params are 1-based but all MIPS docs/tools that I am aware (including IDA) are using a0 - a3 for function params.

2. Pointers got miscalculated. In the original code the pair 

```
lui at, 0x8005
sw a1, 0xFC(at)
```

should be equal to `[(0x8005 << 16) + 0xFC] = a1`, but something got confused with signed/unsigned bit?

3. `0x10000 * a2 - 0x10000;` part is weird while functionally equivalent. I would expect something simpler like `(a2 - 1) << 16;`

**tldr: Desired output**
```
int32_t entry_point(int32_t a0, int32_t a1) {
    *(int32_t *)0x800500FC = (a1 - 1) << 16;
    *(int32_t *)0x800500F8 = a0;
    return 0;
}
```

**Comment by PeterMatula** [comment]
Thank you for the report, this is exactly the kind of stuff we need to make the output better and RetDec more mature.

1. I'm not sure about this. I don't think it would be hard to change it, but I'm also not sure if we will want to. But we will discuss it.
2. Very valid point, it does not make sense to assign to negative pointers. We will fix it.
3. We already do idiom analysis, so this should be just a question of adding a new pattern -- I will investigate.

**Comment by PeterMatula** [comment]
Test binary: [test.zip](https://github.com/avast-tl/retdec/files/2889959/test.zip)
